### PR TITLE
[Dynamic Dashboard] DashboardCard states refactoring.

### PR DESCRIPTION
**Pull request by hafizrahman**
## Description

The previous logic for handling Dashboard Card's availability and display states on the Dashboard and Dashboard Customize screens felt flaky and complicated to me. The main reason was because there's a lot of different states at play, such as:
- whether a card is available to show or not,
- whether an available card can be shown or not on the Dashboard screen,
- whether a card should be shown, or hidden, or shown in "unavailable" state in the Customize screen,
- whether cards and its states are saved locally, (and if yes, what saved states need to be preserved when there's a new update attempt).

**For completeness's sake, here's the availability and visibility logic that the app has to handle:**
1. Each card can be available or not based on various checks (e.g: analytics cards are only available if there are orders, Blaze is available if the site is Blaze-eligible, and so on)
2. "available" here means that a card can be seen in Dashboard and Customize screen.
3. Available cards can be enabled or disabled by users in the Customize screen. An enabled card becomes visible in the Dashboard, and disabled card is invisible. In any case, they're still visible in Customize screen.
4. Special rule: for analytics cards, when they're not available, they are invisible in Dashboard but visible in Customize with the "Unavailable" status.
5. Special rule: for non-analytics cards, when they're not available, they are invisible both in Dashboard and Customize.

### What's in the PR:

This PR attempts to make the handling easier to reason with by:
1.  Updating `DashboardCard` to include new variables and accompanying comments related to availability and display states.
2. This in turns simplifies the UI logic by not having to keep track `unavailableDashboardCards` separately anymore.
3. Updating the logic in `updateDashboardCards()` to make updating cards and preserving setting easier to follow
4. Creating `generateDefaultCards()` to separate concern and make the update function smaller.
5. Updating the UI logic on how to display unavailable/hidden cards.

### Analytics check
1. Start with a site with no orders,
6. Ensure Performance and Top Performers are not shown in Dashboard,
7. Go to Customize, ensure Performance and Top Performers are shown as "unavailable"
8. Go to Orders,
9. Create a new Order,
10. Go back to Dashboard,
11. Ensure Performance and Top Performers are not shown in Dashboard,
12. Go to Customize, ensure Performance and Top Performers are shown as regular cards that can be enabled/disabled and reordered.

### Blaze / Onboarding check
First, either start with a site that doesn't support Blaze, or change the line below in `DashboardViewModel` by hardcoding `canShowBlaze` to false:

```
        let initialCards = generateDefaultCards(canShowOnboarding: canShowOnboarding,
                                                canShowBlaze: canShowBlaze,
                                                canShowAnalytics: canShowAnalytics)
```

then,
1. Start the app,
2. Ensure Blaze is not shown in Dashboard,
3. Go to Customize, ensure Blaze is not shown.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/WooCommerceTests/ViewRelated/Dashboard/DashboardViewModelTests.swift [hunk 1]
```diff
@@ -308,27 +308,7 @@ final class DashboardViewModelTests: XCTestCase {
...
-                            .init(type: .blaze, enabled: true)])
-            }
-        }
-
-        // Then
-        waitUntil {
-            viewModel.dashboardCards.contains { $0.type == .performance && $0.enabled }
-        }
-    }
+    // TODO: Add unit test for initially generated cards, and for the synchronizing between generated and saved cards.
```
Comment (hafizrahman) [inline WooCommerce/WooCommerceTests/ViewRelated/Dashboard/DashboardViewModelTests.swift:311]: Will do this in a separate PR

File: Storage/Storage/Model/DashboardCard.swift [hunk 2]
```diff
@@ -6,12 +6,31 @@ public struct DashboardCard: Codable, Hashable, Equatable, GeneratedCopiable {
     /// The type of dashboard card.
     public let type: CardType
 
-    /// Whether the card is enabled in the Analytics Hub.
+    /// The card's availability state for the site.
+    /// To be set externally based on each card's availability check.
+    public var isAvailable: Bool
```
Comment (itsmeichigo) [inline Storage/Storage/Model/DashboardCard.swift]: We should keep the properties immutable with `let` instead of `var`.

File: Storage/Storage/Model/DashboardCard.swift [hunk 3]
```diff
@@ -6,12 +6,31 @@ public struct DashboardCard: Codable, Hashable, Equatable, GeneratedCopiable {
...
+
+    /// User-changeable setting in the Customize screen, whether to enable or disable an available card.
+    /// An available card will become invisible on the Dashboard, but stay visible on Customize, if `enabled` is set to false.
     public var enabled: Bool
```
Comment (itsmeichigo) [inline Storage/Storage/Model/DashboardCard.swift]: I missed this when writing this file but we should keep this immutable too.

File: Storage/Storage/Model/DashboardCard.swift [hunk 4]
```diff
@@ -6,12 +6,31 @@ public struct DashboardCard: Codable, Hashable, Equatable, GeneratedCopiable {
...
+    var isVisible: Bool {
+        if status == .hide || status == .unavailable {
+            return false
+        } else {
+            return enabled
+        }
+    }
+
+    /// Determines how a card is shown in the Customize screen
+    public var status: CustomizeState
```
Comment (itsmeichigo) [inline Storage/Storage/Model/DashboardCard.swift]: Same comment as above about immutability.

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardView.swift [hunk 5]
```diff
@@ -121,8 +121,8 @@ struct DashboardView: View {
         }
         .sheet(isPresented: $viewModel.showingCustomization) {
             DashboardCustomizationView(viewModel: DashboardCustomizationViewModel(
-                allCards: viewModel.dashboardCards,
-                inactiveCards: viewModel.unavailableDashboardCards,
+                allCards: viewModel.dashboardCards.filter { $0.status != .hide },
+                inactiveCards: viewModel.dashboardCards.filter { $0.status == .unavailable },
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Dashboard/DashboardView.swift]: These are logic and should be moved to the view model for unit testing.

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift [hunk 6]
```diff
@@ -273,46 +269,69 @@ private extension DashboardViewModel {
...
+    func generateDefaultCards(canShowOnboarding: Bool, canShowBlaze: Bool, canShowAnalytics: Bool) -> [DashboardCard] {
+        var cards = [DashboardCard]()
+
+        // Onboarding card.
+        // When not available, Onboarding card needs to be hidden from Dashboard and Customize
+        cards.append(
+            DashboardCard(type: .onboarding,
+                          isAvailable: canShowOnboarding,
+                          enabled: canShowOnboarding,
+                          status: canShowOnboarding ? .show : .hide)
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift]: `isAvailable` and `status` seems duplicated to me. Is it sufficient to use `status` only instead?

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift [hunk 7]
```diff
@@ -273,46 +269,69 @@ private extension DashboardViewModel {
...
+                                                canShowBlaze: canShowBlaze,
+                                                canShowAnalytics: canShowAnalytics)
+
+        // Next, get saved cards and preserve existing enabled state for all available cards.
+        // This is needed because even if a user already disabled an available card and saved it, in `initialCards`
+        // the same card might be set to be enabled. To respect user's setting, we need to check the saved state and re-apply it.
+        let savedCards = await loadDashboardCards() ?? []
+        dashboardCards = initialCards.map { initialCard in
+            if let savedCard = savedCards.first(where: { $0.type == initialCard.type }),
+               savedCard.isAvailable && initialCard.isAvailable {
```
Comment (itsmeichigo) [inline WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift]: Can you clarify the reasoning behind this check?

File: Storage/Storage/Model/DashboardCard.swift (hunk 2 above)
Comment (hafizrahman) [reply Storage/Storage/Model/DashboardCard.swift]: fixed in https://github.com/woocommerce/woocommerce-ios/pull/12626/commits/4d3466adef2a24904643d0b19c73dfdd118f2a78

File: Storage/Storage/Model/DashboardCard.swift (hunk 3 above)
Comment (hafizrahman) [reply Storage/Storage/Model/DashboardCard.swift]: fixed in https://github.com/woocommerce/woocommerce-ios/pull/12626/commits/4d3466adef2a24904643d0b19c73dfdd118f2a78

File: Storage/Storage/Model/DashboardCard.swift (hunk 4 above)
Comment (hafizrahman) [reply Storage/Storage/Model/DashboardCard.swift]: fixed in https://github.com/woocommerce/woocommerce-ios/pull/12626/commits/4d3466adef2a24904643d0b19c73dfdd118f2a78

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardView.swift (hunk 5 above)
Comment (hafizrahman) [reply WooCommerce/Classes/ViewRelated/Dashboard/DashboardView.swift]: Thanks! This is a great point and I moved display logic here and in the model, to the view model, in https://github.com/woocommerce/woocommerce-ios/pull/12626/commits/3fc44de92ea04425d36e625f6aab9e671c124d6a

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift (hunk 6 above)
Comment (hafizrahman) [reply WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift]: Yes! Thanks for noticing this, so glad to have a fresh pair of eyes looking at it. I have been thinking about this for days and it's been so muddled in my mind.

I was able to simplify the code quite a bit following your idea, in https://github.com/woocommerce/woocommerce-ios/pull/12626/commits/3fc44de92ea04425d36e625f6aab9e671c124d6a

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift [hunk 8]
```diff
@@ -273,46 +281,58 @@ private extension DashboardViewModel {
...
+            if let savedCard = savedCards.first(where: { $0.type == initialCard.type }),
+               savedCard.availability == .show && initialCard.availability == .show {
+                return initialCard.copy(enabled: savedCard.enabled)
             } else {
-                // Start with default values, cards could be updated further below as needed.
-                return [DashboardCard(type: .onboarding, enabled: canShowOnboarding),
-                        DashboardCard(type: .performance, enabled: canShowAnalytics),
-                        DashboardCard(type: .topPerformers, enabled: canShowAnalytics),
-                        DashboardCard(type: .blaze, enabled: canShowBlaze)]
+                return initialCard
```
Comment (hafizrahman) [inline WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift:331]: In this function, the logic for updating is:
1. Create set of default cards based on `canShow` parameters,
2. Compare those default cards with the saved cards, and take into account saved card's enabled value if the same card in default and saved are both `.show`

The comparison is needed to deal with this example case:

1. Imagine initially the site has 2 available cards: `onboarding`, `performance`
2. Then, the user sets `onboarding` as enabled, `performance` as disabled, and saves it,
3. Then, the user sets up the site so that Blaze is now available for the site,
4. That triggers `canShowBlaze` value update, which then calls `updateDashboardCards()`
5. Inside the function, `generateDefaultCards()` is then called to create `initialCards`
6. In that array, `performance` is set as enabled (because `canShowAnalytics` remains true as before)
7. Now, we can't set `dashboardCards` with `initialCards` right away, because from point 2 above we know that the user wants `performance` to be disabled.
8. Thus, we need to compare `initialCards` with `savedCards`, find card that's both `.show` in the arrays (indicating they're unchanged after card update), then take the `enabled` state from the saved card.

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift (hunk 7 above)
Comment (hafizrahman) [reply WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift]: I put some explanation for this here: https://github.com/woocommerce/woocommerce-ios/pull/12626/files#r1587606533

File: WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift (hunk 8 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Dashboard/DashboardViewModel.swift:331]: Thanks for the clear explanation, this makes sense!

File: Storage/Storage/Model/DashboardCard.swift [hunk 9]
```diff
@@ -6,11 +6,17 @@ public struct DashboardCard: Codable, Hashable, Equatable, GeneratedCopiable {
...
-    /// Whether the card is enabled in the Analytics Hub.
-    public var enabled: Bool
+    /// User-changeable setting in the Customize screen, whether to enable or disable an available card.
+    /// An available card will become invisible on the Dashboard, but stay visible on Customize, if `enabled` is set to false.
+    public let enabled: Bool
 
-    public init(type: CardType, enabled: Bool) {
+    /// The card's availability state for the site.
+    /// To be set externally based on each card's availability check.
+    public let availability: AvailabilityState
```
Comment (itsmeichigo) [inline Storage/Storage/Model/DashboardCard.swift:15]: I like this new property name, much clearer than `status` from before.

**Review by itsmeichigo — APPROVED** [review]
Thanks for the update @hafizrahman! :shipit:
